### ISSUE TITLE
Log with any `reduce_fx` `Callable`

### ISSUE THREAD

**Issue by jspaezp** [PyTorchLightning/pytorch-lightning#8894]
## 🐛 Bug

Logging fails when passing a callable as a `reduce_fx` with the following exception:

```
pytorch_lightning.utilities.exceptions.MisconfigurationException: Only `self.log(..., reduce_fx={min,max,mean,sum})` are currently supported. Please, open an issue in `https://github.com/PyTorchLightning/pytorch-lightning/issues`. Found: <function nanmean at 0x7fa4b394a9e0>
```

After looking into it, I noticed that the function
https://github.com/PyTorchLightning/pytorch-lightning/blob/e0605472306d6b95bf2616ab88f8c29f4498402e/pytorch_lightning/trainer/connectors/logger_connector/result.py#L89-L104

is typed as accepting a callable but is not implemented to do so.

### To Reproduce

Use `self.log_dict({"loss": loss}, prog_bar = True, reduce_fx = nanmean)`, being `nanmean` a function
https://colab.research.google.com/drive/1XqPbR2AUFWqvT26-HRdoq-k1CD-fJnOh?usp=sharing

### Expected behavior

The expected behavior is that the provided function will be used to reduce the passed arguments and have the result be logged

### Environment

```
* CUDA:
	- GPU:
		- Tesla V100-SXM2-16GB
	- available:         True
	- version:           10.2
* Packages:
	- numpy:             1.19.5
	- pyTorch_debug:     False
	- pyTorch_version:   1.9.0+cu102
	- pytorch-lightning: 1.4.2
	- tqdm:              4.62.0
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.7.11
	- version:           #1 SMP Sat Jun 5 09:50:34 PDT 2021
```

### Posible fix

modify https://github.com/PyTorchLightning/pytorch-lightning/blob/e0605472306d6b95bf2616ab88f8c29f4498402e/pytorch_lightning/trainer/connectors/logger_connector/result.py#L89-L104

```python
        if isinstance(reduce_fx, str):
            reduce_fx = reduce_fx.lower()
            if reduce_fx == "avg":
                reduce_fx = "mean"
            if reduce_fx not in ("min", "max", "mean", "sum"):
                raise MisconfigurationException(error)
            self._reduce_fx = getattr(torch, reduce_fx)
        ##### START ADD
        elif isinstance(reduce_fx, Callable):
            self._reduce_fx = reduce_fx
        ##### END ADD
        elif self.is_custom_reduction:
            raise MisconfigurationException(error)

```

ps: I also noted that the url in the boring model colab notebook, used to download the script that collect environment details is no longer available so it has to be manually modified

**Comment by tchaton** [comment]
Dear @jspaezp,

Would you mind making a PR to resolve both issues ?

**Comment by jspaezp** [comment]
Dear @tchaton 

Would you mind pointing me out how to submit a request for the google colab notebook?
(working on the other one)

Best,
Sebastian

**Comment by carmocca** [comment]
> Only `self.log(..., reduce_fx={min,max,mean,sum})` are currently supported

This is not a bug. The error message is warning that we only support those options.

Support for any `Callable` can be added, but is considered a feature :)

**Comment by jspaezp** [comment]
Ok after a minor internal conversation with @carmocca I think i have come to clarify the issue. 
It seems AS IT IS RIGHT NOW that the problem was clarity in the documentation waaaaay upstream from there the configuration exception was being raised (`trainer.log()` and `trainer.log_dict()`).

```
reduce_fx: reduction function over step values for end of epoch. :meth:`torch.mean` by default.
```

https://github.com/PyTorchLightning/pytorch-lightning/blob/037a86c873dc561db3e6c744514e9550c274f4d6/pytorch_lightning/core/lightning.py#L361

and 

```
    _reduce_fx: Callable = torch.mean
```

Which made it seem like any callable was indeed supported and the check was the problem, turns out its the other way around (only 4 operations are supported, documentation is the problem).

So I think the scope of the solutions changes a bit.

If there is not desired to add support for arbitrary callables, I think the documentation could be improved and type hints should be changed to make clear the intent. Literal("mean" "max", "sum") 

If there is desire to add support for arbitrary callables, after looking more at the code, adding support for it is well beyond the scope of "adding an in statement". (the more I look at this alternative, the less appealing it looks ...)

**Alternatives on the user-side:**
1. Manually add a torchmetric and compute the value before passing it to the logger (so the logger gets the mean of a single value)

**Comment by tchaton** [comment]
Dear @jspaezp,

Would you mind updating the documentation with those ?

Best,
T.C

**Comment by carmocca** [comment]
We could mention in the documentation that this is a limitation and that it can be bypassed using a `torchmetric.Metric`

But we can also add support for arbitrary callables by keeping a list of logged values in memory for those who can afford it.

**Comment by carmocca** [comment]
Update after some offline discussion:

- The main recommendation is to pass a `torchmetrics.Metric` instance for the `reduce_fx` function needed. This avoids duplicating functionality inside the `log` implementation and `torchmetrics`. Also has the benefit that the functionality becomes available to all torchmetrics users.
- If a metric is not in `torchmetrics`, open a feature request for it! You can also write your own.
- There is no plan to add more `reduce_fx=str` options.
- We want to avoid any new flags to `self.log`.
- We could add support for arbitrary `reduce_fx` callables by naively keeping a list of logged values in memory. This could be good to have for test purposes (perhaps trying different things but you don't want to use a `Metric` yet) or those with enough memory to afford it. If this is done, we should print a big warning that this is inefficient compared to passing a `Metric` instance.

**Comment by thesofakillers** [comment]
Please update your docs with this. From the docs, it seems reduce_fx takes any callable, which is not the case. I just spent a lot of time porting my code to pytorch lightning and now I am blocked because of this issue.

**Comment by carmocca** [comment]
Opened a PR with the docs update.

@thesofakillers, what's blocking you exactly? Did you find a `torchmetric.Metric` that covers your use case?
